text/000-rustdoc-cfgs-handling.md

```diff
@@ -0,0 +1,283 @@
+Rustdoc: stabilization of the `doc(cfg*)` attributes
+
+- Features Name: `doc_cfg`
+- Start Date: 2022-12-07
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#43781](https://github.com/rust-lang/rust/issues/43781)
+
+
+# Summary
+[summary]: #summary
+
+This RFC aims at providing rustdoc users the possibility to add visual markers to the rendered documentation to know under which conditions an item is available (currently possible through the following unstable features: `doc_cfg`, `doc_auto_cfg` and `doc_cfg_hide`).
+
+It does not aim to allow having a same item with different `cfg`s to appear more than once in the generated documentation.
+
+It does not aim to document items which are *inactive* under the current configuration (i.e., “`cfg`'ed out”).
+
+# Motivation
+[motivation]: #motivation
+
+The goal of this RFC is to stabilize the possibility to add visual markers to the rendered documentation to know under which conditions an item is available.
+
+Providing this information to users will solve a common issue: “Why can I see this item in the documentation and yet can't use it in my code?”.
+The end goal being to provide this information automatically so that the documentation maintenance cost won't increase.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This RFC proposes to add the following attributes:
+
+  * `#[doc(cfg(...))]`
+
+    This attribute is used to document the operating systems, feature flags, and build profiles where an item is available. For example, `#[doc(cfg(unix))` will add a tag that says "this is supported on **unix** only" to the item.
+
+    The syntax of this attribute is the same as the syntax of the [`#[cfg(unix)]` attribute][cfg attribute] used for conditional compilation.
+
+  * `#[doc(auto_cfg)]`/`#[doc(no_auto_cfg)]`
+
+    When this is turned on, `#[cfg]` attributes are shown in documentation just like `#[doc(cfg)]` attributes are.
+
+  * `#[doc(cfg_hide(...))]` / `#[doc(cfg_show(...))]`
```

Comment:
issue: so i've never been happy with the sheer number of different doc options this feature introduces: the names are kind of all over the place and hard to predict.

I think `#[doc(cfg(..))]` is fine, but it would be nice if the rest all derived from each other.

Perhaps `#[doc(auto_cfg(disable))]`, `#[doc(auto_cfg(show(...), hide(...))]` etc? Ideally we only introduced _one_ new doc attribute but I'm not sure if that's easy since `cfg` can accept anything. Potentially could use the distinction between `cfg(..)` and `cfg = ...` here. Thoughts?

Reply:
I personally wouldn't like using `cfg = ...` since it would just make the distinction more confusing. I'm not really sure there's a massive benefit to having `doc(auto_cfg(show(...)))` versus `doc(cfg_show(...))` since, while it is literally combining two attributes into one, it just feels like unnecessary extra typing to me.

That said, perhaps there would be a case for having `doc(auto_cfg(...))` for `doc(cfg_show(...))` and `doc(auto_cfg(not(...)))` for `doc(cfg_hide(...))`, since it would fit in with existing `cfg` and be a bit more discoverable. Perhaps `doc(auto_cfg(all))` and `doc(auto_cfg(none))` could be allowed as aliases for enabling/disabling all options.

Still has some issues, but, if we were to change the existing attributes, I'd prefer something like that over just combining the `show` and `hide` into a single `auto_cfg` attribute.

Reply:
@clarfonthey point taken on subtle syntax, I didn't really want to propose that anyway which is why I proposed it coming out of `auto_cfg`

To me the main distinction between `cfg_show` and `auto_cfg(show)` is the latter is easier to remember: there are many potential names for `cfg_show` and I won't remember which one it is. Especially since `auto_cfg` has `cfg` as the _second_ word. It's inconsistent and confusing.

If everything trees out of a single attribute it's easier to remember IMO, and not confusing.

Reply:
And yes I'm happy with experimentation with `auto_cfg`, I didn't want to propose one particular solution there just want it to be something straightforward, intuitive, and easy to remember.

Reply:
I'm not exactly convinced that `auto_cfg(show(...))` is easier to remember compared to `cfg_show(...)`, and in both cases, IDEs can offer suggestions after typing `doc(cfg` to let you know what's available.

My main point here is that having "fewer attributes" isn't really a principle we should be guided by, since what matters most is the user writing the code, not the compiler parsing it. To a user, `auto_cfg(show(...))` and `auto_cfg(hide(...))` are still two attributes; they just require extra parentheses compared to `cfg_show(...)` and `cfg_hide(...)`.

Reply:
I do not think "everyone is using an IDE" is either correct or a driving design principle we should have.

My argument here is entirely from the user perspective. I find these names confusing and hard to remember, and a logical grouping would simplify it. I do not care one whit how easy or hard it is for the compiler to parse.

Reply:
I'm also open to that though I still prefer having fewer attributes

Reply:
I agree with the notion that an IDE should not be assumed; that was tangential to the point that adding extra parentheses doesn't really affect the mental model of how to do things here, IMHO. Show and hide are a pair, whether they are rested inside an "auto" or not.

Reply:
Right, but they _do_ apply only to autocfg, it's part of what makes them confusing.

Reply:
Another option could be `#[doc(auto_cfg_hide(...))]` then.

Though, I'm now wondering whether there's a way to change the semantics slightly so that `cfg_hide` can take over the role of `no_auto_cfg` too. Instead of disabling `auto_cfg` completely just tell it to hide _everything_ so it doesn't actually do anything. (Combined with my discussion above of restricting this to the crate-root that means there would only be two new attributes required in this RFC: `#[doc(cfg(...))]` anywhere and `#![doc(cfg_hide(...))]` at the root).

Reply:
I don't understand what you suggested @Nemo157. Do you mean that if no argument is provided to `cfg_hide`, it means that we hide everything?

So to summarize the discussion, it seems that we tend to for `show` and `hide`:

```rust
#[doc(cfg_show(test))]
#[doc(cfg_hide(test))]
```

As for enabling/disabling the feature, what do you think about:

```rust
#[doc(cfg_auto(enable)]
#[doc(cfg_auto(disable)]
```

?

Reply:
My understanding of the issue, after completely misunderstanding for a while (apologies, @Manishearth) is that the `cfg_show` and `cfg_hide` attributes make no reference to `auto_cfg` and so it's not clear that they control *automatically* showing and hiding `cfg` attributes. Manish' proposal of `auto_cfg(show(...))` and `auto_cfg(hide(...))` is one way of accomplishing this, but `auto_cfg_show` and `auto_cfg_hide` attributes could also accomplish this.

Reply:
> I don't understand what you suggested @Nemo157. Do you mean that if no argument is provided to `cfg_hide`, it means that we hide everything?

I don't have a concrete syntax proposal, I just wanted to seed the idea in case someone else has thoughts on how it could be accomplished. Being able to reduce this feature to introducing a total of two attributes instead of five I think is a helpful simplification.

Reply:
@clarfonthey yes, thanks. my two main complaints are that the split of attributes starting and ending with cfg is confusing (and the potential for them to do either makes it harder to remember _even if_ we pick consistent names), and the second one is the one you describe: it's not clear that show/hide is about auto stuff.